Add Cpp Version of Matmul Benchmark

edwinleeTT opened the pull request:
### Ticket
#18269

### Problem description
A C++ version of the existing Python matmul benchmark test was required.

### What's changed
This adds test_benchmark.cpp, which is a TTNN adaptation of tt-metal/tests/ttnn/unit_tests/benchmarks/test_benchmark.py. Changes are also made to the test fixture to enable tracing. Like the Python version, this test is skipped by default to prevent running in pipelines.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI [passes](https://github.com/tenstorrent/tt-metal/actions/runs/14089410972)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI [passes](https://github.com/tenstorrent/tt-metal/actions/runs/14095775590)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

edwinleeTT committed:
Move to new branch - adds test with functional tracing. Issues in utilistaion results reported.

edwinleeTT committed:
Corrects metric gathering and increases allocated space for tracing.

edwinleeTT committed:
Adds scoping for profiler runs

edwinleeTT committed:
Add skip to ensure test isn't run in CI

edwinleeTT committed:
Update program cache assertion to check that a warmup run is used

edwinleeTT committed:
Add SPDX copyright

omilyutin-tt commented on tests/ttnn/unit_tests/gtests/ttnn_test_fixtures.hpp:
```diff
@@ -27,6 +27,10 @@ using namespace tt::tt_metal;  // For test
 
 class TTNNFixtureWithDevice : public ::testing::Test {
+private:
+    int trace_region_size;
+    int l1_small_size;
```
* Add suffix `_` to clearly disambiguate these variables are class data members
* Initialize the primitive types to some default value; e.g. `= DEFAULT_TRACE_REGION_SIZE;` or `= DEFAULT_L1_SMALL_SIZE` in this case. https://github.com/tenstorrent/tt-metal/blob/4910164b2d7860b810121c02f5d1f42b345c1f39/contributing/BestPractices.md#15-initialize-primitive-types-on-declaration

omilyutin-tt commented on tests/ttnn/unit_tests/gtests/test_benchmark.cpp:
This file should probably be named in a way that indicates the benchmarks are for matmul? Are you adding more benchmarks here?

edwinleeTT replied:
Right, this makes sense. I've renamed the file to make it a bit clearer. The name was to match the original version of this test, [test_benchmark.py](https://github.com/tenstorrent/tt-metal/blob/main/tests/ttnn/unit_tests/benchmarks/test_benchmark.py)

omilyutin-tt commented on tests/ttnn/unit_tests/gtests/test_benchmark.cpp:
```diff
@@ -0,0 +1,412 @@
+};
+
+class Matmul2DHostPerfTestFixture : public ttnn::TTNNFixtureWithDevice,
+                                    public testing::WithParamInterface<std::tuple<
+                                        /* grid_size */ std::tuple<int, int>,
+                                        /* tile_h */ int,
+                                        /* tile_w */ int,
+                                        /* num_warmup_iterations */ int,
+                                        /* num_measurement_iterations */ int,
+                                        /* use_program_cache */ bool>> {
```
The test should be parameterized over `matmul_configs` and shapes - this is what `TEST_P` does for you. Scope out params that are unchanged in the test body itself.

edwinleeTT replied:
Test is reworked - hopefully it is more suitable now

omilyutin-tt commented on tests/ttnn/unit_tests/gtests/test_benchmark.cpp:
```diff
@@ -0,0 +1,412 @@
+    return {1, 1};
+}
+
+std::vector<std::tuple<int, int, int, bool, bool, int, int, int>> matmul_shapes_bfloat16 = {
```
Here and in other places in the PR - don't use `std::tuple` / `std::pair`. Create `struct` with named fields for tuples and use `Shape2D` for pairs.

edwinleeTT replied:
Done!

ayerofieiev-tt reviewed (commented):
Thank you for working on this! Customers asked for a C++ version of the benchmark for a while now, so its great to see this happening now.

ayerofieiev-tt commented on tests/ttnn/unit_tests/gtests/test_benchmark.cpp:
```diff
@@ -0,0 +1,412 @@
+    {16384, 16384, 16384, false, false, 4, 4, 4},
+};
+
+std::vector<std::tuple<DataType, MathFidelity, bool>> matmul_configs = {
```
what bool means here?

edwinleeTT replied:
I was whether tracing was enabled - this has been rewritten so hopefully it's much clearer now

bbradelTT commented on tests/ttnn/unit_tests/gtests/test_benchmark.cpp:
```diff
@@ -0,0 +1,412 @@
+
+            const ttnn::operations::matmul::Matmul matmul_params(
+                program_config,
+                /*bcast_batch*/ std::nullopt,
```
Please use the format
```
/*bcast_batch=*/std::nullopt
```
here and elsewhere

bbradelTT commented on tests/ttnn/unit_tests/gtests/test_benchmark.cpp:
```diff
@@ -0,0 +1,412 @@
+            std::vector<float> in1_data(k * n);
+            std::generate(in1_data.begin(), in1_data.end(), []() {
+                float value = static_cast<float>(rand()) / static_cast<float>(RAND_MAX);
+                return std::round(value * 15.0f) / 15.0f;
```
Why 15.0f?
Please create a well named constant.

edwinleeTT replied:
I will delete this, I was just doing some experiments with the inputs and missed cleaning this up. It isn't needed to maintain the original functionality of inputs from [0,1).

ayerofieiev-tt commented on tests/ttnn/unit_tests/gtests/test_benchmark.cpp:
```diff
@@ -0,0 +1,412 @@
+                in0_data,
+                ttnn::TensorSpec(
+                    ttnn::Shape({m, k}),
+                    tt::tt_metal::TensorLayout(dtype, tt::tt_metal::Layout::TILE, in0_memory_config)),
```
If you tell that Layout is `TILE`, it expects `in0_data` to be tiled already. 
It kind of does not matter when in0_data is all ones, except maybe cases for bfloat8b_ and bfloat4_b

bbradelTT commented on tests/ttnn/unit_tests/gtests/test_benchmark.cpp:
```diff
@@ -0,0 +1,412 @@
+                /* out_subblock_w */ out_subblock_w,
+                /* out_block_h */ out_block_h,
+                /* out_block_w */ out_block_w,
+                /* per_core_M */ per_core_M,
```
You should remove redundant comments like this, while keeping the ones that are not redundant (e.g. 
```
/*transpose_mcast=*/false,
/*fused_activation=*/std::nullopt
```

ayerofieiev-tt commented on tests/ttnn/unit_tests/gtests/test_benchmark.cpp:
```diff
@@ -0,0 +1,412 @@
+                    ttnn::Shape({k, n}),
+                    tt::tt_metal::TensorLayout(dtype, tt::tt_metal::Layout::TILE, ttnn::DRAM_MEMORY_CONFIG)),
+                device);
+            ttnn::operations::matmul::MatmulMultiCoreReuseMultiCastProgramConfig program_config{
```
afaik, there are many other config types.
can you leave info on when to use one or another?

edwinleeTT replied:
My understanding is that the only performant configs we should use are the 1D, 2D, and DRAM configs. I will add a few words to the benchmark here explaining this. 

There is also some information available [here](https://github.com/tenstorrent/tt-metal/blob/7c36ba9ecc00bc98dffd4cebae680d3eb071afef/tech_reports/ViT-TTNN/vit.md) about the matmul configs.

ayerofieiev-tt commented on tests/ttnn/unit_tests/gtests/test_benchmark.cpp:
```diff
@@ -0,0 +1,412 @@
+                /* transpose_mcast */ false,
+                /* fused_activation */ std::nullopt};
+
+            const ttnn::WormholeComputeKernelConfig compute_kernel_config =
```
does it mean this won't work on Blackhole?
CC @bbradelTT

bbradelTT replied:
It should work on Blackhole. For the compute kernel config, it was decided that both Wormhole and Blackhole would share the same Wormhole compute kernel config.

I don't know what will happen fur future architectures, although I would expect they will all use WormholeComputeKernelConfig until new features are added that require a new class.

ayerofieiev-tt replied:
can we remove GraySkullComputeKernelConfig and rename WormholeComputeKernelConfig to ComputeKernelConfig?

Can we give it such a name thats its going to be clear that its legit for both archs?

bbradelTT replied:
That would be too intrusive of a change, and it's all supposed to be abstracted away behind
ttnn/cpp/ttnn/operations/core/compute_kernel/compute_kernel_config.hpp

@edwinleeTT please change the code to be along the lines of
```
ttnn::DeviceComputeKernelConfig compute_kernel_config = ttnn::init_device_compute_kernel_config(...);
```

edwinleeTT replied:
I validated the`WormholeComputeKernelConfig` can be used on Blackhole, before removing it in favour of using `init_device_compute_kernel_config` as suggested above (which is also validated on both WH + BH)

ayerofieiev-tt commented on tests/ttnn/unit_tests/gtests/test_benchmark.cpp:
```diff
@@ -0,0 +1,412 @@
+            // Warmup iterations
+            for (int iter = 0; iter < num_warmup_iterations; ++iter) {
+                output_tensor = ttnn::operations::matmul::matmul(
+                    in0_t,
+                    in1_t,
```
very nit, minor, but 
`_t` here is for `tensor`?
overlaps with `_t` in types like `uint32_t`

bbradelTT commented on tests/ttnn/unit_tests/gtests/test_benchmark.cpp:
```diff
@@ -0,0 +1,412 @@
+            std::string utilization_user_grid_percentage = std::to_string(utilization_user_grid * 100);
+            tt::log_info(
+                "M*K*N = {}*{}*{} == inference time (avg): {}, tflops (avg): {}, utilization (vs user grid): {}%, "
+                "utilization (vs 8x8 grid): {}%",
```
Please put a note in about this being normalized and point to the python file.

bbradelTT replied:
Maybe try using `tt::Cluster::instance().get_device_aiclk(device->id());` and see if that gets you the frequency?

edwinleeTT replied:
Thanks for the suggestion - this worked for both Blackhole and Wormhole

ayerofieiev-tt commented on tests/ttnn/unit_tests/gtests/test_benchmark.cpp:
```diff
@@ -0,0 +1,412 @@
+                    in1_t,
+                    /* bias */ std::nullopt,
+                    /* parameters */ matmul_params);
+                device->push_work(
+                    [device]() mutable { Synchronize(device, std::nullopt, std::vector<SubDeviceId>()); });
+                device->synchronize();
```
why do you need to call both?

edwinleeTT replied:
I believe this ensures the main and worker thread both complete, this was also to match the Python implementation via the [Pybind](https://github.com/tenstorrent/tt-metal/blob/978c711569bc23a04e448fe52bdd6181d076ba40/ttnn/cpp/pybind11/device.cpp#L571)

bbradelTT commented on tests/ttnn/unit_tests/gtests/test_benchmark.cpp:
```diff
@@ -0,0 +1,412 @@
+            } else {
+                auto start_time = std::chrono::high_resolution_clock::now();
+                {
+                    ZoneScopedN("Matmul iterations");
```
What do the different code paths look like with profiling?

bbradelTT commented:
What are the results of trying this and the python version on BH?

bbradelTT commented on tests/ttnn/unit_tests/gtests/test_benchmark.cpp:
```diff
@@ -0,0 +1,412 @@
+                        in1_t,
+                        /* bias */ std::nullopt,
+                        /* parameters */ matmul_params);
+                    output_tensor.deallocate();
```
@edwinleeTT why do we need to have deallocate tensor inside the loop while in tests/ttnn/unit_tests/benchmarks/test_benchmark.py it was at the end of the benchmark?

And where are the output tensors allocated (L1 vs DRAM)?

edwinleeTT replied:
My understanding is that the Python version deallocates data upon it going out of scope. This means that tensors didn't need to be explicitly deallocated in the original Python reference, while in the C++ version it is needed to achieve the same behaviour (and avoid a memory leak caused by allocating a new output tensor every iteration of the loop). The output tensors will either be in L1 or DRAM, depending on the test configuration, and both are exercised when running the full benchmark. 

If this explicit allocation is removed, the larger matmuls that output to L1 will exceed the available L1 and cause a test failure

bbradelTT replied:
Ok. Thanks.

bbradelTT commented on tests/ttnn/unit_tests/gtests/test_benchmark.cpp:
```diff
@@ -0,0 +1,412 @@
+                        device->push_work(
+                            [device]() mutable { Synchronize(device, std::nullopt, std::vector<SubDeviceId>()); });
+                        device->synchronize();
+                        output_tensor.deallocate();
```
Ideally, you can move the deallocate to the end of the benchmark, then you can also move the synchronize outside of the loop and before end_time.

If you can't move the deallocate to later on, please test out what happens if you only measure the matmul+synchronize time instead of matmul+synchronize+deallocate.

edwinleeTT replied:
The deallocate can't be moved to the end of the benchmark unless the benchmark is constrained to only run on small matmul configurations that won't run out of L1 space with this leak. 

Omitting the deallocate from timing does not result in a significant difference in measured performance (1-2%).

bbradelTT replied:
Thanks.

edwinleeTT commented on tests/ttnn/unit_tests/gtests/test_benchmark.cpp:
```diff
@@ -0,0 +1,412 @@
+                out_sharded ? ttnn::MemoryConfig{ttnn::TensorMemoryLayout::BLOCK_SHARDED, ttnn::BufferType::L1}
+                            : ttnn::DRAM_MEMORY_CONFIG;
+
+            const Tile output_tile =
+                out_sharded ? (tile_h <= 16 ? tt::tt_metal::Tile({tile_h, 32}) : tt::tt_metal::Tile({tile_h, tile_w}))
+                            : tt::tt_metal::Tile({tile_h, tile_w});
```
Ah, I copied the logic exactly from the [reference](https://github.com/tenstorrent/tt-metal/blob/e725458329c4205d71fd7d8252f61122cce33bbf/tests/ttnn/unit_tests/benchmarks/test_benchmark.py#L295) - I will simplify it a bit.

edwinleeTT committed:
Add default initialisation and underscore suffix to member variables

edwinleeTT committed:
PR Feedback:

-Replaces tuples and pairs, and uses gtest to iterate over the test configs.
-More clearly label the paramaters for functions
-Use the more abstract init_device_compute_kernel_config
-Explain the different matmul config options
-Simplify dimension calculation slightly

edwinleeTT committed:
Rename test to more clearly indicate the purpose of running and timing a series of matmuls

edwinleeTT committed:
Rename to have leading "test" matching other tests

edwinleeTT committed:
Remove hardcoded device frequency in favour of using get_device_aiclk

bbradelTT commented on tests/ttnn/unit_tests/gtests/test_matmul_benchmark.cpp:
```diff
@@ -0,0 +1,814 @@
+    const bool use_trace = test_config.enable_tracing_;
+
+    TT_FATAL(num_measurement_iterations > 0, "Won't have data without at least one measurement iteration");
+    TT_ASSERT(
```
Please either remove or turn into a TT_FATAL

edwinleeTT replied:
Removed, in case there is a desire to run without warming up the program cache for some reason.

omilyutin-tt commented on tests/ttnn/unit_tests/gtests/test_matmul_benchmark.cpp:
```diff
@@ -0,0 +1,814 @@
+    return {1, 1};
+}
+
+struct Matmul_Test_Config {
```
Remove `_` here, follow `CamelCase` for C++ types

omilyutin-tt commented on tests/ttnn/unit_tests/gtests/test_matmul_benchmark.cpp:
```diff
@@ -0,0 +1,814 @@
+}
+
+struct Matmul_Test_Config {
+    DataType dtype_ = DataType::BFLOAT16;
```
Can drop `_` suffix for structs, also no need fr constructor. You can rely on the default generated

edwinleeTT replied:
Changed and removed

omilyutin-tt commented on tests/ttnn/unit_tests/gtests/test_matmul_benchmark.cpp:
```diff
@@ -0,0 +1,814 @@
+        num_measurement_iterations_(num_measurement_iterations) {}
+};
+
+struct Matmul_Shape {
```
Same comments as above

omilyutin-tt commented on tests/ttnn/unit_tests/gtests/test_matmul_benchmark.cpp:
```diff
@@ -0,0 +1,814 @@
+class Matmul2DHostPerfTestFixture : public ttnn::TTNNFixtureWithDevice,
+                                    public testing::WithParamInterface<std::tuple<Matmul_Test_Config, Matmul_Shape>> {
+public:
+    Matmul2DHostPerfTestFixture() : ttnn::TTNNFixtureWithDevice(65536, 200000) {}
```
nit: can you annotate what these numbers mean with `/*var_name=*/ 65536`?

edwinleeTT replied:
Good catch, added!

omilyutin-tt commented on tests/ttnn/unit_tests/gtests/test_matmul_benchmark.cpp:
```diff
@@ -0,0 +1,814 @@
+
+    // Parse test config
+    const Matmul_Test_Config& test_config = std::get<0>(GetParam());
+    const int& num_warmup_iterations = test_config.num_warmup_iterations_;
+    const int& num_measurement_iterations = test_config.num_measurement_iterations_;
+    const bool& use_program_cache = test_config.use_program_cache_;
```
It is more efficient to use by-value semantics for primitive types, drop `&` here

omilyutin-tt commented on tests/ttnn/unit_tests/gtests/test_matmul_benchmark.cpp:
```diff
@@ -0,0 +1,814 @@
+    const Matmul_Shape& matmul_shape = std::get<1>(GetParam());
+
+    const Shape2D& grid_size = matmul_shape.grid_size_;
+    const int& tile_h = matmul_shape.tile_shape_.height();
+    const int& tile_w = matmul_shape.tile_shape_.width();
```
Same here

omilyutin-tt commented on tests/ttnn/unit_tests/gtests/test_matmul_benchmark.cpp:
```diff
@@ -0,0 +1,814 @@
+            {DataType::BFLOAT16, MathFidelity::HiFi4, /*enable_tracing=*/true}}),
+        // Matmul_Shapes to use
+        ::testing::ValuesIn(std::vector<Matmul_Shape>{
+            {/*m=*/512,
+             /*k=*/512,
+             /*n=*/512,
+             /*in0_sharded=*/true,
+             /*out_sharded=*/true,
+             /*in0_block_w_div=*/1,
+             /*num_out_blocks_h=*/1,
+             /*num_out_blocks_w=*/1},
```
This is nice, but you can rely on aggregate initialization `.m = ..., .k = ..., ` etc

omilyutin-tt commented on tests/ttnn/unit_tests/gtests/ttnn_test_fixtures.hpp:
```diff
@@ -36,10 +40,20 @@ class TTNNFixtureWithDevice : public ::testing::Test {
 
     void TearDown() override { tt::tt_metal::CloseDevice(device_); }
+
+public:
```
👍

omilyutin-tt commented on tests/ttnn/unit_tests/gtests/test_matmul_benchmark.cpp:
```diff
@@ -0,0 +1,814 @@
+
+    TT_FATAL(grid_size.height() > 0 && grid_size.width() > 0, "Invalid grid size");
+
+    const char* TT_METAL_HOME = std::getenv("TT_METAL_HOME");
```
Use `snake_case` for variables

omilyutin-tt commented on tests/ttnn/unit_tests/gtests/test_matmul_benchmark.cpp:
```diff
@@ -0,0 +1,814 @@
+    return oss.str();
+}
+
+std::vector<Shape2D> SUBBLOCK_HW_CHOICES = {{4, 2}, {2, 4}, {8, 1}, {1, 8}, {7, 1}, {1, 7}, {3, 2},
```
`std::vector` has non-trivial dtor, can we do `std::array` here or wrap in `Indestructible`  https://github.com/tenstorrent/tt-metal/blob/4910164b2d7860b810121c02f5d1f42b345c1f39/contributing/BestPractices.md#17-avoid-static-variables-with-non-trivial-destructors

Also `kSubblockHwChoices` I think is a bit more common, and more common in the industry in general

omilyutin-tt commented on tests/ttnn/unit_tests/gtests/test_matmul_benchmark.cpp:
```diff
@@ -0,0 +1,814 @@
+
+    const ttnn::operations::matmul::Matmul matmul_params(
+        program_config,
+        /*bcast_batch=*/std::nullopt,
```
👍

omilyutin-tt commented on tests/ttnn/unit_tests/gtests/test_matmul_benchmark.cpp:
```diff
@@ -0,0 +1,814 @@
+             /*num_out_blocks_w=*/8}})));
+
+INSTANTIATE_TEST_SUITE_P(
+    /*Prefix for the instantiated tests*/ MatmulTests_BFLOAT4B,
+    /*Test suite*/ Matmul2DHostPerfTestFixture,
```
I think it is fine to drop here

The main intention for the annotation is to distinguish unnamed ints / booleans like you did below for `Matmul_Shape`

edwinleeTT committed:
PR Feedback - cleaning up the structs, naming conventions, static memory usage.

edwinleeTT committed:
Make variables consistent snake_case, move fidelity cycle count into own function

edwinleeTT committed:
Include tt_cluster for clang tidy check
